main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -0,0 +1,214 @@
+import ca.uwaterloo.flix.util.collection.CofiniteSet
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+
+import java.util.concurrent.ConcurrentLinkedQueue
+
+/**
+  * The inliner optionally performs beta-reduction at call-sites.
+  */
+object Inliner {
+
+  /**
+    * Performs inlining on the given AST `root`.
+    */
+  def run(root: OccurrenceAst.Root)(implicit flix: Flix): OccurrenceAst.Root = {
+    implicit val sctx: SharedContext = SharedContext.mk()
+    val defs = ParOps.parMapValues(root.defs)(visitDef(_)(root, sctx, flix))
+    val newRoot = OccurrenceAst.Root(defs, root.enums, root.structs, root.effects, root.mainEntryPoint, root.entryPoints, root.sources)
+    newRoot
+  }
+
+  /**
+    * Performs expression inlining on the given definition `def0`.
+    * Converts definition from [[OccurrenceAst]] to [[OccurrenceAst]].
+    */
+  private def visitDef(def0: OccurrenceAst.Def)(implicit root: OccurrenceAst.Root, sctx: SharedContext, flix: Flix): OccurrenceAst.Def = def0 match {
+    case OccurrenceAst.Def(sym, fparams, spec, exp, ctx, loc) =>
+      val e = visitExp(exp, Context.Empty)(sym, root, sctx, flix)
+      OccurrenceAst.Def(sym, fparams, spec, e, ctx, loc)
+  }
+
+  /**
+    * Performs inlining operations on the expression `exp0` from [[Expr]].
+    * Returns a [[Expr]]
+    */
+  private def visitExp(exp0: Expr, ctx0: Context)(implicit sym0: Symbol.DefnSym, root: OccurrenceAst.Root, sctx: SharedContext, flix: Flix): Expr = exp0
+
+  /**
+    * Returns `true` if `def0` should be inlined.
+    */
+  private def canInlineDef(defCtx: DefContext, ctx: Context): Boolean = {
+    val mayInline = !defCtx.isSelfRecursive && !ctx.currentlyInlining
+    val shouldInline = defCtx.isDirectCall
+    mayInline && shouldInline
+  }
+
+  private def isPure(eff0: Type): Boolean = {
+    eff0 == Type.Pure
+  }
+
+  /**
+    * Checks if `occur` is Dead.
+    */
+  private def isDead(occur: OccurrenceAst.Occur): Boolean = occur match {
+    case Dead => true
+    case _ => false
+  }
+
+  /**
+    * Checks if `occur` is Once and `eff` is Pure
+    */
+  private def isUsedOnceAndPure(occur: OccurrenceAst.Occur, eff0: Type): Boolean = occur match {
+    case Once => isPure(eff0)
+    case _ => false
+  }
+
+  /**
+    * Checks if `exp0` is trivial and `eff` is pure
+    */
+  private def isTrivialAndPure(exp0: Expr, eff0: Type): Boolean = {
+    isPure(eff0) && isTrivialExp(exp0)
+  }
+
+  /**
+    * Checks if `occur` is dead and `exp` is pure.
+    */
+  private def isDeadAndPure(occur: OccurrenceAst.Occur, eff0: Type): Boolean = occur match {
+    case Dead => isPure(eff0)
+    case _ => false
+  }
+
+  /** Returns a canonical effect type equivalent to `eff` */
+  private def canonicalEffect(eff: Type): Type = {
+    evalToType(eval(eff), eff.loc)
+  }
+
+  /** Evaluates a ground, simplified effect type */
+  private def eval(eff: Type): CofiniteSet[Symbol.EffectSym] = eff match {
```

Comment:
Want example

Reply:
Maybe @JonathanStarup can help with this. This is something I don't quite fully understand since the previous inliner used `Purity` instead of `Type` for effects so comparison on purity was simply checking ` x == Purity.Pure`